File: app-area-agent-cti/src/main/java/com/lsxy/app/area/cti/busnetcli/Callbacks.java
package com.lsxy.app.area.cti.busnetcli;

/**
 * BUS 客户端的回调函数集合接口类
 *
 * @author 雪彦
 */
public interface Callbacks {

    /**
     * 全局BUS连接状态变化回调
     *  @param unitId     产生连接状态变化的BUS节点的Unit ID
     * @param clientId   产生连接状态变化的BUS节点的Client ID
     * @param clientType 产生连接状态变化的BUS节点的Client Type
     * @param status     产生连接状态变化的BUS节点的状态
     * @param addInfo    产生连接状态变化的BUS节点的附加信息
     */
    public void globalConnect(byte unitId, byte clientId, byte clientType, byte status, String addInfo);

    /**
     * 连接结果回调
     *  @param localClientId     连接结果产生变化的本地客户端的ID
     * @param accessPointUnitId 连接结果产生变化的本地客户端所连接的服务器的 UnitID
     * @param errorCode         错误编码。链接失败时返回。0表示成功
     */
    public void connect(byte localClientId, int accessPointUnitId, int errorCode);

    /**
     * 断线回调
     *
     * @param localClientId 连接断开的本地客户端的ID
     */
    public void disconnect(byte localClientId);

    /**
     * 收到数据回调
     *  @param head 数据收发者等信息
     * @param bytes 二进制数据
     */
    public void data(Head head, byte[] bytes);

    /**
     * 收到日志输出的回调函数
     *
     * @param msg   日志内容
     * @param isErr 是否错误日志
     */
    public void log(String msg, Boolean isErr);
}

File: busnetcli-jni/src/callbacks.cpp
#include "callbacks.h"
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <jni.h>
#include "globvars.h"

static JNIEnv* get_current_thread_env() {
    JNIEnv* env = NULL;
    if (jvm->GetEnv((void**) (&env), JNI_VERSION_1_8) == JNI_EDETACHED)
        jvm->AttachCurrentThread((void**) (&env), NULL);
    return env;
}

void connection(void * arg, unsigned char local_client_id,
                int accesspoint_unit_id, int ack) {
    JNIEnv* env = get_current_thread_env();
    env->CallStaticVoidMethod(
        cls_client, meth_client_connect, (jbyte) local_client_id, (jint) accesspoint_unit_id, (jint) ack
    );
}

void disconnect(void * arg, unsigned char local_client_id) {
    JNIEnv* env = get_current_thread_env();
    env->CallStaticVoidMethod(cls_client, meth_client_connect, (jbyte) local_client_id);
}

void recvdata(void * arg, unsigned char local_client_id,
              SMARTBUS_PACKET_HEAD * head, void * data, int size) {
    JNIEnv* env = get_current_thread_env();
    jbyteArray jbytes_data = env->NewByteArray(size);
    env->SetByteArrayRegion(jbytes_data, 0, size, data);
    env->CallStaticVoidMethod(
        cls_client, meth_client_recvdata,
        (jbyte) head->cmd, (jbyte) head->cmdtype,
        (jbyte) head->src_unit_id, (jbyte) head->src_unit_client_id, (jbyte) head->src_unit_client_type,
        (jbyte) head->dest_unit_id, (jbyte) head->dest_unit_client_id, (jbyte) head->dest_unit_client_type,
        jbytes_data
    );
}

void global_connect(void * arg, char unit_id, char client_id, char client_type,
                    char accesspoint_unit, char status, const char * add_info) {
    JNIEnv* env = get_current_thread_env();
    jstring txt = env->NewStringUTF(add_info);
    env->CallStaticVoidMethod(
        cls_client, meth_client_globalconenct,
        (jbyte) unit_id, (jbyte) client_id, (jbyte) client_type,
        (jbyte) status, txt
    );
}

void trace(const char * msg) {
    if (!msg)
        return;
    if (!msg[0])
        return;
    size_t msg_sz = strlen(msg);
    if (!msg_sz)
        return;
    JNIEnv* env = get_current_thread_env();
    jbyteArray jbytes_msg = env->NewByteArray(msg_sz);
    env->SetByteArrayRegion(jbytes_msg, 0, msg_sz, msg);
    env->CallStaticVoidMethod(
        cls_client, meth_client_log,
        jbytes_msg, (jboolean) false
    );
}

void trace_err(const char * msg) {
    if (!msg)
        return;
    if (!msg[0])
        return;
    size_t msg_sz = strlen(msg);
    if (!msg_sz)
        return;
    JNIEnv* env = get_current_thread_env();
    jbyteArray jbytes_msg = env->NewByteArray(msg_sz);
    env->SetByteArrayRegion(jbytes_msg, 0, msg_sz, msg);
    env->CallStaticVoidMethod(
        cls_client, meth_client_log,
        jbytes_msg, (jboolean) true
    );
}
